### PR TITLE
VMware - Secure Boot & Virtual TPM

### PR DESCRIPTION
**Pull request by stejskalleos**
Add support for Secure Boot and TPM
Inspired by https://github.com/ananace/foreman_vmware_advanced
Required FOG PR: https://github.com/fog/fog-vsphere/pull/305

### CODE REVIEW

File: app/views/compute_resources_vms/form/vmware/_base.html.erb [hunk 1]
```diff
@@ -49,6 +49,16 @@ end %>
                { :disabled => images.empty?, :label => _('Image'), :label_size => "col-md-2" } %>
 </div>
 
+<%= checkbox_f f, :secure_boot, { :help_inline => _("Enable Secure Bott for provisioning."),
```
Comment (chris1984) [inline app/views/compute_resources_vms/form/vmware/_base.html.erb]:
```suggestion
<%= checkbox_f f, :secure_boot, { :help_inline => _("Enable Secure Boot for provisioning."),
```

**Review by chris1984 — COMMENTED** [review]
One comment, looks like tests are failing since we don't have fog-vsphere built with your change.

File: app/views/compute_resources_vms/form/vmware/_base.html.erb (hunk 1 above)
Comment (nofaralfasi) [inline app/views/compute_resources_vms/form/vmware/_base.html.erb]: I think we should disable this option if the firmware is set to BIOS (as we did with Libvirt).

File: app/models/compute_resources/foreman/model/vmware.rb [hunk 2]
```diff
@@ -761,6 +761,22 @@ def normalize_vm_attrs(vm_attrs)
...
+      attrs[:secure_boot] = ActiveRecord::Type::Boolean.new.cast(enabled)
+    end
+
+    def virtual_tpm
+      attrs[:virtual_tpm] ||= false
+    end
+
+    def virtual_tpm=(enabled)
+      attrs[:virtual_tpm] = ActiveRecord::Type::Boolean.new.cast(enabled)
+    end
```
Comment (nofaralfasi) [inline app/models/compute_resources/foreman/model/vmware.rb]: Why is this needed? I couldn't find the use of it.

**Review by ekohl — CHANGES_REQUESTED** [review]
With the firmware we have a field `auto`. Wouldn't it make more sense to add a value there for UEFI + SecureBoot? The default is `auto`, which derives it from the PXE loader. We should enhance that to automatically detect if secureboot was desired.

**Comment by nofaralfasi** [comment]
> With the firmware we have a field `auto`. Wouldn't it make more sense to add a value there for UEFI + SecureBoot? The default is `auto`, which derives it from the PXE loader. We should enhance that to automatically detect if secureboot was desired.

I didn't know that's how the Automatic option works. I'll have to adjust my recent changes accordingly.

**Comment by ekohl** [comment]
It works by deriving the `firmware_type` from the `pxe_loader`:
https://github.com/theforeman/foreman/blob/7ce3336d67e1b0b825b8a4add92a340c7d2d144b/app/models/concerns/pxe_loader_support.rb#L49-L59

This is then used in the `firmware_mapping`:
https://github.com/theforeman/foreman/blob/7ce3336d67e1b0b825b8a4add92a340c7d2d144b/app/models/compute_resources/foreman/model/vmware.rb#L818-L821

And in parsing the arguments this is used to determine the `firmware`:
https://github.com/theforeman/foreman/blob/7ce3336d67e1b0b825b8a4add92a340c7d2d144b/app/models/compute_resources/foreman/model/vmware.rb#L495

In my libvirt EFI patch I copied this pattern:
https://github.com/theforeman/foreman/pull/9965/files#diff-525a301145c328b1c4d2e6dd7668a9ce4b9f64edb7b2876a4cdebeb5ee2689f9R162-R177

You can see I already added a TODO there to include secure boot.

This is completely lost in https://github.com/theforeman/foreman/pull/10209.

**Comment by nofaralfasi** [comment]
Requires: https://github.com/fog/fog-vsphere/pull/305

**Comment by nofaralfasi** [comment]
UI changes in my commit:
- Added a new firmware type `UEFI Secure Boot` for Secure Boot.
- Removed the Secure Boot checkbox (replaced it with the new firmware type).
- Added a label helper indicating that TPM is only compatible with EFI firmware.
- Hid the TPM option from the UI when it is not relevant.

image

File: app/models/compute_resources/foreman/model/vmware.rb [hunk 3]
```diff
@@ -206,6 +206,7 @@ def firmware_types
         "automatic" => N_("Automatic"),
         "bios" => N_("BIOS"),
         "efi" => N_("EFI"),
+        "uefi_sb" => N_("UEFI Secure Boot"),
       }
     end
```
Comment (nofaralfasi) [inline app/models/compute_resources/foreman/model/vmware.rb]: Consider moving this method to the `ComputeResource` model to be used by both VMware and Libvirt.

File: app/models/compute_resources/foreman/model/vmware.rb [hunk 4]
```diff
@@ -816,8 +822,14 @@ def vm_instance_defaults
...
-      'bios'
+      case firmware_type
+      when :uefi
+        'efi'
+      when :uefi_sb
+        'uefi_sb'
+      else
+        'bios'
+      end
     end
```
Comment (nofaralfasi) [inline app/models/compute_resources/foreman/model/vmware.rb]: Same here, consider moving this method to the `ComputeResource` model to be used by both VMware and Libvirt.

File: app/views/compute_resources_vms/form/vmware/_base.html.erb [hunk 5]
```diff
@@ -49,6 +50,14 @@ end %>
                { :disabled => images.empty?, :label => _('Image'), :label_size => "col-md-2" } %>
 </div>
 
+<!-- # BIOS firmware doesn't support virtual TPM. -->
+<div id='efi_features' <%= 'class=hide' if !new_vm || f.object.firmware == 'bios' %>>
```
Comment (ekohl) [inline app/views/compute_resources_vms/form/vmware/_base.html.erb]: Do I get it right you can add and remove the virtual TPM module from an existing VM?

One workflow I wonder about: assume the user selected a BIOS PXE loader (which now results in a BIOS firmware) and also selects the TPM module. What happens?

**Review by ekohl — CHANGES_REQUESTED** [review]
Test failure looks related. I think you need to delete (and then handle) `virtual_tpm` in `parse_args`.

File: app/models/compute_resources/foreman/model/vmware.rb (hunk 3 above)
Comment (ekohl) [reply app/models/compute_resources/foreman/model/vmware.rb]: Sounds like a good idea, but I'd separate it from this PR just to keep it focused.

File: app/views/compute_resources_vms/form/vmware/_base.html.erb (hunk 5 above)
Comment (nofaralfasi) [reply app/views/compute_resources_vms/form/vmware/_base.html.erb]: The opposite. We hide this option if it is an existing VM or if the firmware is set to BIOS. 
However, there's an edge case where the firmware is set to Automatic, then the TPM option is available, and if you select it and set the PXE Loader to BIOS, you will get the following error:
`Failed to power up a compute VMware_with_keys (VMware) instance <HOST NAME>: GenericVmConfigFault: The virtual machine must use EFI firmware.`

**Comment by nofaralfasi** [comment]
> Test failure looks related. I think you need to delete (and then handle) `virtual_tpm` in `parse_args`.

I thought it's because we're missing the https://github.com/fog/fog-vsphere/pull/305 PR.

File: app/views/compute_resources_vms/form/vmware/_base.html.erb (hunk 5 above)
Comment (ekohl) [reply app/views/compute_resources_vms/form/vmware/_base.html.erb]: > The opposite. We hide this option if it is an existing VM or if the firmware is set to BIOS.

`f.object.firmware` can only be set on an existing VM. If you always hide it on an existing VM, isn't it redundant? In other words:
```suggestion
<div id='efi_features' <%= 'class=hide' if !new_vm %>>
```

> However, there's an edge case where the firmware is set to Automatic, then the TPM option is available, and if you select it and set the PXE Loader to BIOS, you will get the following error:

This is what I was getting at. At first, I'd catch the situation in code before it's even sent to VMware so you can raise an error in the form. Then you can decide if you want to add some front end code to handle this. It may be useful given the two form controls are on different tabs but I wouldn't block on it when there's a proper handling post-submit.

**Comment by ekohl** [comment]
No, it's because it tries to assign it to the host model now instead of transforming it to a vsphere api parameter

**Comment by chris1984** [comment]
@nofaralfasi did you want to address @ekohl concerns before I test, that way we can test it just once and get this in faster once the code looks good.

**Comment by ekohl** [comment]
> concerns before I test, that way we can test it just once and get this in faster once the code looks good.

I assure you, you'll see the exact same failures as the automated tests so there's no point in testing manually.

File: app/views/compute_resources_vms/form/vmware/_base.html.erb (hunk 5 above)
Comment (nofaralfasi) [reply app/views/compute_resources_vms/form/vmware/_base.html.erb]: > f.object.firmware can only be set on an existing VM. If you always hide it on an existing VM, isn't it redundant?

Actually, I was trying to address another edge case scenario where the user chooses the BIOS firmware. If they encounter an error when clicking submit, the firmware will be set to BIOS, and we don't want to display the TPM option.

I will handle this error on the backend as well.

**Comment by nofaralfasi** [comment]
> No, it's because it tries to assign it to the host model now instead of transforming it to a vsphere api parameter

Where do you see it in the tests result?  I still think it's because of the missing PR from fog-vsphere.

File: app/views/compute_resources_vms/form/vmware/_base.html.erb (hunk 5 above)
Comment (nofaralfasi) [reply app/views/compute_resources_vms/form/vmware/_base.html.erb]: @chris1984 I wouldn't block testing because of it. Users should know that the TPM option is not available with BIOS. I think this is an edge case (the user would have to set the firmware type to automatic, the PXE loader to BIOS, and enable the TPM option), which would be nice to cover but shouldn't block us from testing.

**Comment by ekohl** [comment]
Oh yes, I think I misread that. Apologies for that. Perhaps you can modify the PR to actually point to the PR so we can see if the tests would pass otherwise.

**Comment by nofaralfasi** [comment]
I have another idea on how to implement this. Moving it to WIP until it's ready.

**Comment by nofaralfasi** [comment]
This is the error displayed to the user if they enable the TPM option while their firmware is set to BIOS:
image

File: app/models/compute_resources/foreman/model/vmware.rb (hunk 3 above)
Comment (nofaralfasi) [reply app/models/compute_resources/foreman/model/vmware.rb]: After working on this feature for both Libvirt and VMware, I believe it makes sense to move this method and the `firmware_mapping` method to the ComputeResource model for shared use.

**Comment by nofaralfasi** [comment]
Changes:
- TPM option is always visible on the host's edit page (disabled for existing hosts). 
- Handles TPM conflict with BIOS. This is the error displayed to the user if they enable the TPM option while their firmware is set to BIOS:
image

Note: tests are in progress.

File: app/models/compute_resources/foreman/model/vmware.rb [hunk 6]
```diff
@@ -491,6 +483,18 @@ def parse_args(args)
...
       args[:firmware] = firmware_mapping(firmware_type) if args[:firmware] == 'automatic'
 
+      virtual_tpm = args.delete(:virtual_tpm)
+      args[:virtual_tpm] = ActiveRecord::Type::Boolean.new.cast(virtual_tpm) unless virtual_tpm.nil?
+
+      if args[:firmware] == 'bios' && args[:virtual_tpm]
+        errors.add :base, _('TPM is not compatible with BIOS firmware. Please change Firmware or disable TPM.')
+      end
+
+      if args[:firmware] == 'uefi_sb'
```
Comment (sbernhard) [inline app/models/compute_resources/foreman/model/vmware.rb]: Isn't this something, which can be handled in the `firmware_mapping` method, too?

File: app/models/compute_resource.rb [hunk 7]
```diff
@@ -399,6 +399,26 @@ def normalize_vm_attrs(vm_attrs)
...
+      "uefi_sb" => N_("UEFI Secure Boot"),
+    }.freeze
+  end
+
+  def firmware_mapping(firmware_type)
+    case firmware_type
+    when :uefi
+      'efi'
+    when :uefi_sb
+      'uefi_sb'
```
Comment (ekohl) [inline app/models/compute_resource.rb]: I'd avoid abbreviations and write it out in full
```suggestion
    when :uefi_secure_boot
      'uefi_secure_boot'
```

File: app/models/compute_resources/foreman/model/vmware.rb (hunk 6 above)
Comment (ekohl) [reply app/models/compute_resources/foreman/model/vmware.rb]: This is the translation layer of Compute Resource to actual fog-vsphere API. Today that isn't unified. It would be interesting if fog-vsphere (and fog-libvirt) understood the same parameters so code is unified.

File: app/models/compute_resources/foreman/model/vmware.rb [hunk 8]
```diff
@@ -523,6 +527,7 @@ def create_vm(args = { })
       else
         vm = new_vm(args)
         vm.firmware = 'bios' if vm.firmware == 'automatic'
+        raise errors.full_messages.join(', ') unless errors.empty?
```
Comment (ekohl) [inline app/models/compute_resources/foreman/model/vmware.rb]: What can add `errors` here? We already have `return unless errors.empty?` a bit earlier and besides `parse_networks` I don't think there are things that should errors. Also it's inconsistent that this raises while the other simply returns.

**Review by ekohl — CHANGES_REQUESTED** [review]
If I'm pedantic, I'd expect to see the move of `firmware_types` and `firmware_mapping` in its own commit. The later on add the additional values. That makes it a bit easier for the reviewer to see what's going on. I'd also question if it should be part of this PR. Making it part of the one where it'll be shared with libvirt is IMHO the more logical place.

Then you can also think about what other compute resources should do with it. I think `firmware_mapping` is questionable in its current form, because with `uefi_sb` it doesn't really map to a single value anymore. Or really, taking a step back: if you try to summarize `firmware_mapping`, what would it be? What's the `firmware_type` parameter?

File: app/models/compute_resources/foreman/model/vmware.rb (hunk 8 above)
Comment (nofaralfasi) [reply app/models/compute_resources/foreman/model/vmware.rb]: The first check `errors.empty?` is for the test_connection. However, in this case, we're checking if any errors occurred while calling the new_vm method. If the firmware is set to BIOS and the virtual TPM is set to true, we will get an error: https://github.com/theforeman/foreman/blob/0c60e217451551ed27ba87a9b33178200101e7bc/app/models/compute_resources/foreman/model/vmware.rb#L490

I chose this approach to avoid disrupting the entire process. If the user is using the UI, I want to retain the form with the values they have set and only display this specific error message.

**Comment by ekohl** [comment]
@nofaralfasi and I had some discussion and I came up with https://github.com/ekohl/foreman/commit/707162487891fdd9d2d744c5f52a83470c9d0d72 as some possible improvements.

**Comment by nofaralfasi** [comment]
Explanation of the Update:
Creating a new CR machine differs from creating a bare-metal machine, which means we currently lack a way to validate the VM form fields submitted by the user.

The recent updates have introduced a scenario where setting the firmware type to BIOS and enabling the new Virtual TPM option results in an error, as TPM is not compatible with BIOS.
To address this, the best solution would be to catch this error before sending the data to fog-vsphere and raise an error for the user, while preserving the form with its current values.

Future Improvements:
This part of the code would benefit from future enhancements, such as providing better descriptions of the fields (e.g., explaining what it means to set the firmware to Automatic) and adding validation of the data before it is sent to fog-vsphere. Unfortunately, we lack this functionality not only for the parameters added in this PR but also for the existing ones. If we want to refactor it, we should do it all together, which would require a larger PR.

File: app/models/compute_resources/foreman/model/vmware.rb [hunk 9]
```diff
@@ -488,8 +497,22 @@ def parse_args(args)
...
+          args[:firmware] = firmware_type
+        end
+      end
+
+      if args[:firmware]&.start_with?('uefi')
+        args[:secure_boot] = true if args[:firmware] == 'uefi_secure_boot'
+        args[:firmware] = 'efi'
+      end
+
+      args[:virtual_tpm] = ActiveRecord::Type::Boolean.new.cast(args.delete(:virtual_tpm))
```
Comment (ekohl) [inline app/models/compute_resources/foreman/model/vmware.rb]: No need to delete it:
```suggestion
      args[:virtual_tpm] = ActiveRecord::Type::Boolean.new.cast(args[:virtual_tpm])
```

I'd still insist that we raise an exception here. That way `clone_vm` (which also calls parse_args) can benefit from it as well.

When we earlier discussed adding an error, it was important to also add a return if there are any errors (like `create_vm` already does).

The benefit of `errors` is that you can actually flag the field itself that is wrong, instead of some generic unspecified exception.

File: app/models/compute_resources/foreman/model/vmware.rb [hunk 10]
```diff
@@ -207,7 +207,16 @@ def firmware_types
         "automatic" => N_("Automatic"),
         "bios" => N_("BIOS"),
         "efi" => N_("EFI"),
-      }
+        "uefi_secure_boot" => N_("UEFI Secure Boot"),
```
Comment (ekohl) [inline app/models/compute_resources/foreman/model/vmware.rb:210]: I think it's inconsistent to name this `uefi_secure_boot` while the other option is `efi`. Should we add `uefi` as well and deprecate `efi`?

File: app/views/compute_resources_vms/form/vmware/_base.html.erb [hunk 11]
```diff
@@ -49,6 +51,12 @@ end %>
                { :disabled => images.empty?, :label => _('Image'), :label_size => "col-md-2" } %>
 </div>
 
+<%= checkbox_f f, :virtual_tpm, { :help_inline => _("Add Virtual TPM module to the VM."),
+                                  :label => _('Virtual TPM'),
+                                  :label_help => _("Only compatible with EFI firmware."),
```
Comment (ekohl) [inline app/views/compute_resources_vms/form/vmware/_base.html.erb]: Here it will also be confusing if we have both EFI and UEFI. A user may assume that it's incompatible with UEFI Secure Boot while in practice it's not.

You may solve it like this:
```suggestion
                                  :label_help => _("Only compatible with (U)EFI firmware."),
```

But I'd argue to standardize everywhere on UEFI. Wikipedia also redirects from EFI to UEFI.

File: test/models/compute_resources/vmware_test.rb [hunk 12]
```diff
@@ -398,13 +398,27 @@ class Foreman::Model::VmwareTest < ActiveSupport::TestCase
...
+
       test 'chooses BIOS firmware when no pxe loader is set and firmware is automatic' do
         attrs_in = HashWithIndifferentAccess.new('firmware' => 'automatic')
         attrs_out = {:firmware => "bios"}
         assert_equal attrs_out, @cr.parse_args(attrs_in)
       end
     end
 
+    context 'virtual_tpm' do
+      test 'sets virtual_tpm to true when firmware is EFI and virtual_tpm is enabled' do
```
Comment (ekohl) [inline test/models/compute_resources/vmware_test.rb]: I'd also expect test cases that cover BIOS firmware and UEFI secure boot

File: bundler.d/vmware.rb [hunk 13]
```diff
@@ -1,3 +1,5 @@
 group :vmware do
-  gem 'fog-vsphere', '>= 3.6.4', '< 4.0'
+  # gem 'fog-vsphere', '>= 3.6.4', '< 4.0'
+  # TODO: Remove this once the fog-vsphere gem is released, and upgrade to the latest version
+  gem 'fog-vsphere', :git => "https://github.com/fog/fog-vsphere.git", :branch => "master"
```
Comment (ekohl) [inline bundler.d/vmware.rb]: Looks like 3.6.8 was released which includes your code.

On a related note: looks like this was also missed in the NVME commit.

**Review by ekohl — CHANGES_REQUESTED** [review]

File: app/models/compute_resources/foreman/model/vmware.rb (hunk 10 above)
Comment (nofaralfasi) [reply app/models/compute_resources/foreman/model/vmware.rb:210]: I'm concerned that this might confuse users, as VMware refers to EFI but not UEFI.

Comment (nofaralfasi) [reply app/models/compute_resources/foreman/model/vmware.rb:210]: I must admit that I'm also confused. After reviewing VMware’s documentation, I see that they refer to both EFI and UEFI firmware. However, in their UI, only EFI is listed as a firmware type.

File: app/views/compute_resources_vms/form/vmware/_base.html.erb (hunk 11 above)
Comment (nofaralfasi) [reply app/views/compute_resources_vms/form/vmware/_base.html.erb]: If you think that’s clearer, I’ll proceed with the change.

File: app/models/compute_resources/foreman/model/vmware.rb (hunk 9 above)
Comment (nofaralfasi) [reply app/models/compute_resources/foreman/model/vmware.rb]: I believe `clone_vm` shouldn’t encounter this issue. Based on my review, it also calls the `create_vm` method later in the process, and the workflow is consistent with that for creating a new VM.

Regarding adding this as an error, it might not be very effective since we would still end up initializing the Fog::Vsphere::Compute::Server object [here](https://github.com/theforeman/foreman/blob/develop/app/models/compute_resources/foreman/model/vmware.rb#L552).

IMO, the correct place to raise this exception is in the `parse_args` method. However, since we don't have the validations for this, this would end up in a 500 error code, causing the whole page to break. 
Unfortunately, I couldn't find a better solution for this (that doesn't require refactoring this entire part of the code).
As I mentioned in my previous comment, we lack this functionality for all the fields sent to Fog::Vsphere. Addressing it for just this specific case will not be sufficient.

File: test/models/compute_resources/vmware_test.rb (hunk 12 above)
Comment (nofaralfasi) [reply test/models/compute_resources/vmware_test.rb]: I need to test the `create_vm` method. Testing only `parse_args` is not sufficient.

**Comment by nofaralfasi** [comment]
In progress: improve testing for `virtual_tpm`.

File: app/models/compute_resources/foreman/model/vmware.rb (hunk 10 above)
Comment (ekohl) [reply app/models/compute_resources/foreman/model/vmware.rb:210]: https://en.wikipedia.org/wiki/UEFI uses them rather interchangeable. https://docs.vmware.com/en/VMware-vSphere/7.0/com.vmware.vsphere.vm_admin.doc/GUID-898217D4-689D-4EB5-866C-888353FE241C.html only uses UEFI. Have you looked how it's shown in the vSphere UI?

File: app/views/compute_resources_vms/form/vmware/_base.html.erb (hunk 11 above)
Comment (ekohl) [reply app/views/compute_resources_vms/form/vmware/_base.html.erb]: I'd still prefer to align everything on 1 term so it's clear it's the same under the hood, just secure boot or not.

File: app/models/compute_resources/foreman/model/vmware.rb (hunk 10 above)
Comment (nofaralfasi) [reply app/models/compute_resources/foreman/model/vmware.rb:210]: Yeah, in the vSphere UI they use EFI, not UEFI.

File: app/models/compute_resources/foreman/model/vmware.rb [hunk 14]
```diff
@@ -488,8 +497,22 @@ def parse_args(args)
...
+          args[:firmware] = firmware_type
+        end
+      end
+
+      if args[:firmware]&.start_with?('uefi')
+        args[:secure_boot] = true if args[:firmware] == 'uefi_secure_boot'
+        args[:firmware] = 'efi'
+      end
+
+      args[:virtual_tpm] = ActiveRecord::Type::Boolean.new.cast(args[:virtual_tpm])
```
Comment (ekohl) [inline app/models/compute_resources/foreman/model/vmware.rb:515]: I still think we can raise an exception here:
```suggestion
      args[:virtual_tpm] = ActiveRecord::Type::Boolean.new.cast(args[:virtual_tpm])
      if args[:virtual_tpm] && args[:firmware] != 'efi'
        raise ArgumentError, _('TPM is incompatible with BIOS firmware. Change firmware or disable TPM.')
      end
```

File: app/models/compute_resources/foreman/model/vmware.rb [hunk 15]
```diff
@@ -488,8 +497,22 @@ def parse_args(args)
...
+
+      if args[:firmware] == 'automatic'
+        if firmware_type == 'none' || firmware_type.empty?
+          args[:firmware] = 'bios'
+        else
+          args[:firmware] = firmware_type
+        end
+      end
+
+      if args[:firmware]&.start_with?('uefi')
```
Comment (sbernhard) [inline app/models/compute_resources/foreman/model/vmware.rb]: For me, this is pretty hard to understand why the firmware type is changed from uefi to efi. Maybe a comment why this needs to be done would help for the furture.
Additionally, I think the code can be simplified:

```
args[:firmware] = 'efi' if args[:firmware]&.start_with?('uefi')
args[:secure_boot] = args[:firmware] == 'uefi_secure_boot'
```

File: app/views/compute_resources_vms/form/vmware/_base.html.erb [hunk 16]
```diff
@@ -6,9 +6,11 @@
   <%= counter_f(f, :corespersocket, label: _('Cores per socket'), recommended_max_value: compute_resource.max_cpu_count, value: f.object.corespersocket || 1) %>
 </div>
 <%= text_f f, :memory_mb, :class => "col-md-2", :label => _("Memory (MB)") %>
+
+<% firmware_type = compute_resource.firmware_type(f.object) %>
 <%= field(f, :firmware, :label => _('Firmware'), :label_size => "col-md-2") do
   compute_resource.firmware_types.collect do |type, name|
-    radio_button_f f, :firmware, {:disabled => !new_vm, :value => type, :text => _(name)}
+    radio_button_f f, :firmware, {:checked => (firmware_type == type), :disabled => !new_vm, :value => type, :text => _(name) }
```
Comment (ekohl) [inline app/views/compute_resources_vms/form/vmware/_base.html.erb]: This is me being pedantic, but previously the whitespace around the braces was consistent and now it's no longer the case:
```suggestion
    radio_button_f f, :firmware, {:checked => (firmware_type == type), :disabled => !new_vm, :value => type, :text => _(name)}
```

**Review by ekohl — CHANGES_REQUESTED** [review]
I agree the current code isn't great at relaying information.

What happens now is the orchestration calls `setCompute`:
https://github.com/theforeman/foreman/blob/2f947f7f31f960f8b297a764d4b4d070b1b250f8/app/models/concerns/orchestration/compute.rb#L98-L118

This calls the `create_vm` which calls `parse_args`. This means you can't really use the errors framework to get specific field level errors and only can raise an exception. Worse, we're actually performing expensive API calls (in `parse_network`) when we could fail early.

Still, `parse_args` can raise an exception and that should bubble up. You can then also add tests without having to test the whole `create_vm` method.

My thought process behind it is that you should fail as early as possible. Longer term I'd suggest that we should refactor `setCompute` to call something like `normalize_args` that effectively does that `parse_args` does now and returns early. Before any connections to the real compute resource have been made.

File: app/models/compute_resources/foreman/model/vmware.rb [hunk 17]
```diff
@@ -488,8 +497,22 @@ def parse_args(args)
 
       args.except!(:hardware_version) if args[:hardware_version] == 'Default'
 
-      firmware_type = args.delete(:firmware_type)
-      args[:firmware] = firmware_mapping(firmware_type) if args[:firmware] == 'automatic'
+      firmware_type = args.delete(:firmware_type).to_s
+
+      if args[:firmware] == 'automatic'
+        if firmware_type == 'none' || firmware_type.empty?
```
Comment (sbernhard) [inline app/models/compute_resources/foreman/model/vmware.rb]:
```
args[:firmware] = if firmware_type == 'none' || firmware_type.empty?
                              'bios'
                            else
                              firmware_type
                            end
```

File: app/models/compute_resources/foreman/model/vmware.rb [hunk 18]
```diff
@@ -206,8 +206,17 @@ def firmware_types
...
         "automatic" => N_("Automatic"),
         "bios" => N_("BIOS"),
-        "efi" => N_("EFI"),
-      }
+        "uefi" => N_("UEFI"),
+        "uefi_secure_boot" => N_("UEFI Secure Boot"),
+      }.freeze
+    end
+
+    def firmware_type(vm_obj)
```
Comment (sbernhard) [inline app/models/compute_resources/foreman/model/vmware.rb:215]:
```
return vm_obj.firmware unless vm_obj.firmware == 'efi'
vm_obj.secure_boot ? "uefi_secure_boot" : "uefi"
```

Comment (ekohl) [reply app/models/compute_resources/foreman/model/vmware.rb:215]: Please no, I think that's distinctly worse to follow. The current flow is very easy to read IMHO.

File: app/models/compute_resources/foreman/model/vmware.rb (hunk 17 above)
Comment (ekohl) [reply app/models/compute_resources/foreman/model/vmware.rb]: Then I'd suggest:
```ruby
args[:firmware] = (firmware_type == 'none' || firmware_type.empty?) ? 'bios' : firmware_type
```

File: app/models/compute_resources/foreman/model/vmware.rb (hunk 15 above)
Comment (ekohl) [reply app/models/compute_resources/foreman/model/vmware.rb]: I think that comment should describe the block above (automatic firmware) as well. The handling as a whole is not intuitive so a comment can make sense.

**Comment by nofaralfasi** [comment]
> This calls the `create_vm` which calls `parse_args`. This means you can't really use the errors framework to get specific field level errors and only can raise an exception. Worse, we're actually performing expensive API calls (in `parse_network`) when we could fail early.

Before that, we are calling the new_vm method also from here [compute_object](https://github.com/theforeman/foreman/blob/develop/app/models/concerns/orchestration/compute.rb#L37). So if we raise the error from parse_args, the exception will result in breaking the whole page. 

**Update:**
I can catch the exception in the compute_object method, which will break the form and display an error message to the user. I was trying to avoid this because I consider it an edge case and didn't want an error that breaks the whole form. However, this approach ensures that the exception is presented in a user-friendly manner, and it's been handled before sending the values to VMware.

> My thought process behind it is that you should fail as early as possible...

I agree, but as I mentioned before, I propose adding validations for all the values before we send them to VMware.

**Comment by ekohl** [comment]
> Before that, we are calling the new_vm method also from here [compute_object](https://github.com/theforeman/foreman/blob/develop/app/models/concerns/orchestration/compute.rb#L37). So if we raise the error from parse_args, the exception will result in breaking the whole page.

Ah, this is something I missed altogether. Then I think your approach is good enough now and we need do something more to untangle the mess.

These are the reasons the host creation/edit form is so complex and why so many attempts to improve it end up failing.

File: app/models/compute_resources/foreman/model/vmware.rb [hunk 19]
```diff
@@ -488,8 +498,24 @@ def parse_args(args)
...
+      end
+
+      # Adjust firmware and secure_boot values for VMware compatibility
+      args[:secure_boot] = true if args[:firmware] == 'uefi_secure_boot'
+      # VMware expects the firmware type to be 'efi'
+      args[:firmware] = 'efi' if args[:firmware]&.start_with?('uefi')
+
+      args[:virtual_tpm] = ActiveRecord::Type::Boolean.new.cast(args[:virtual_tpm])
+      if args[:firmware] == 'bios' && args[:virtual_tpm]
+        raise ArgumentError, _('TPM is not compatible with BIOS firmware. Please change Firmware or disable TPM.')
```
Comment (ekohl) [inline app/models/compute_resources/foreman/model/vmware.rb]: Given what you pointed out, perhaps we can go back to `errors.add(:virtual_tpm, _('TPM is not compatible with BIOS firmware. Please change Firmware or disable TPM.'))`

Then later (I'll add a comment there too) we can check for errors

File: app/models/compute_resources/foreman/model/vmware.rb [hunk 20]
```diff
@@ -522,7 +548,6 @@ def create_vm(args = { })
         clone_vm(args)
       else
         vm = new_vm(args)
```
Comment (ekohl) [inline app/models/compute_resources/foreman/model/vmware.rb:550]: Then you can raise the error here (probably with some message):
```suggestion
        vm = new_vm(args)
        raise ArgumentError if errors.any?
```

**Review by ekohl — CHANGES_REQUESTED** [review]
Given what you told me, I think the previous version with the raising in `create_vm` was indeed better. I'd prefer to revert back to that, or try what I suggest here in this review. I don't know if this works, but my untested theory is that the error should show up in the form field with TPM itself.

**Comment by nofaralfasi** [comment]
This PR needs to be updated to align with the changes in [foreman#10321](https://github.com/theforeman/foreman/pull/10321).

**Comment by nofaralfasi** [comment]
Closing this PR in favor of a new one: https://github.com/theforeman/foreman/pull/10324. The new PR addresses the same issue with updated changes. Please continue the discussion and review there.
